lib/Sema/TypeCheckDecl.cpp

```diff
@@ -3996,6 +3996,125 @@ static void finalizeAbstractStorageDecl(TypeChecker &TC,
   }
 }
 
+static void adjustFunctionTypeForOverride(Type &type) {
+  // Drop 'throws'.
+  // FIXME: Do we want to allow overriding a function returning a value
+  // with one returning Never?
+  auto fnType = type->castTo<AnyFunctionType>();
+  auto extInfo = fnType->getExtInfo();
+  extInfo = extInfo.withThrows(false);
+  if (fnType->getExtInfo() != extInfo)
+    type = fnType->withExtInfo(extInfo);
+}
+
+/// Drop the optionality of the result type of the given function type.
+static Type dropResultOptionality(Type type, unsigned uncurryLevel) {
+  // We've hit the result type.
+  if (uncurryLevel == 0) {
+    if (auto objectTy = type->getOptionalObjectType())
+      return objectTy;
+
+    return type;
+  }
+
+  // Determine the input and result types of this function.
+  auto fnType = type->castTo<AnyFunctionType>();
+  auto parameters = fnType->getParams();
+  Type resultType =
+      dropResultOptionality(fnType->getResult(), uncurryLevel - 1);
+
+  // Produce the resulting function type.
+  if (auto genericFn = dyn_cast<GenericFunctionType>(fnType)) {
+    return GenericFunctionType::get(genericFn->getGenericSignature(),
+                                    parameters, resultType,
+                                    fnType->getExtInfo());
+  }
+
+  return FunctionType::get(parameters, resultType, fnType->getExtInfo());
+}
+
+static Type getMemberTypeForComparison(ASTContext &ctx, ValueDecl *member,
+                                       ValueDecl *derivedDecl = nullptr,
+                                       bool stripLabels = true) {
+  auto *method = dyn_cast<AbstractFunctionDecl>(member);
+  ConstructorDecl *ctor = nullptr;
+  if (method)
+    ctor = dyn_cast<ConstructorDecl>(method);
+
+  auto abstractStorage = dyn_cast<AbstractStorageDecl>(member);
+  assert((method || abstractStorage) && "Not a method or abstractStorage?");
+  SubscriptDecl *subscript = nullptr;
+  if (abstractStorage)
+    subscript = dyn_cast<SubscriptDecl>(abstractStorage);
+
+  auto memberType = member->getInterfaceType();
+  if (derivedDecl) {
+    auto *dc = derivedDecl->getDeclContext();
+    auto owningType = dc->getDeclaredInterfaceType();
+    assert(owningType);
+
+    memberType = owningType->adjustSuperclassMemberDeclType(member, derivedDecl,
+                                                            memberType);
+    if (memberType->hasError())
+      return memberType;
+  }
+
+  if (stripLabels)
+    memberType = memberType->getUnlabeledType(ctx);
+
+  if (method) {
+    // For methods, strip off the 'Self' type.
+    memberType = memberType->castTo<AnyFunctionType>()->getResult();
+    adjustFunctionTypeForOverride(memberType);
+  }
+
+  if (subscript) {
```

Comment:
Should this be an 'else if'? Otherwise you run the '// For properties...' block on methods too.

Reply:
Original code had a `if` in there I just moved it up, but I think you are right, it should be `else if`.